harness/determined/experimental/inference/_torch_batch_process.py

```diff
@@ -0,0 +1,405 @@
+        )
+        """
+        return self._tensorboard_path
+
+    def get_device(self):
+        """
+        Get default device associated with this worker
+        """
+        return self._device
+
+    def get_distributed_context(self):
+        return self._core_context.distributed
+
+    def prepare_model_for_inference(self, model: nn.Module) -> nn.Module:
+        model.eval()
+        model.to(self._device)
+        return model
+
+    def get_default_storage_path(self) -> Iterator[pathlib.Path]:
+        """
+        Should use with "with" syntax
+        """
+        self._use_default_storage = True
+        return self._core_context.checkpoint._storage_manager.store_path(self._storage_path)
+
+
+def get_default_device(core_context: core.Context) -> torch.device:
+    local_rank = core_context.distributed.local_rank
+    local_num_gpu = torch.cuda.device_count()
+    if local_rank >= local_num_gpu:
+        return torch.device("cpu")
+    else:
+        return torch.device("cuda", local_rank)
+
+
+def initialize_distributed_backend() -> Optional[core.DistributedContext]:
+    info = det.get_cluster_info()
+
+    distributed_backend = det._DistributedBackend()
+    if distributed_backend.use_torch():
+        if torch.cuda.is_available():
+            dist.init_process_group(backend="nccl")  # type: ignore
+        else:
+            dist.init_process_group(backend="gloo")  # type: ignore
+        return core.DistributedContext.from_torch_distributed()
+    elif info and (len(info.container_addrs) > 1 or len(info.slot_ids) > 1):
+        raise ValueError(
+            "In multi-slot managed cluster training, you must wrap your training script with a "
+            "distributed launch layer such as determined.launch.torch_distributed"
+        )
+    return None
+
+
+def initialize_default_inference_context() -> core.Context:
+    distributed_context = initialize_distributed_backend()
+    # Use WorkerAskChief mode to ensure synchronize correctly across worker
+    # Using WorkerAskMaster mode could lead to some workers exiting when others
+    # are waiting for synchronization.
+    return det.core.init(
+        distributed=distributed_context, preempt_mode=core.PreemptMode.WorkersAskChief
+    )
+
+
+class TorchBatchProcessor(metaclass=abc.ABCMeta):
+    """
+    User can initialize necessary resources in the init function, such as
+    - model for prediction
+    - storage client (e.g. s3 client)
+    """
+
+    @abc.abstractmethod
+    def __init__(self, context: TorchBatchProcessorContext) -> None:
+        pass
+
+    @abc.abstractmethod
+    def process_batch(self, batch, batch_idx) -> None:
+        pass
+
+    def run_before_checkpoint(self) -> None:
+        """
+        Overwrite this function to run certain logic before checkpointing
+        This function will be called right before each checkpoint.
+        """
+        pass
+
+    def clean_up(self) -> None:
```

Comment:
`on_finish`? or something to signal the "callback-like" nature of this method and the above method.